docs/RFCS/20171118_role_based_access_control.md

```diff
@@ -0,0 +1,747 @@
+* **indirect member**: A is an indirect member of B if `A ∈ C ... ∈ B` where `...` is be an arbitrary number of memberships
+
+## User-level explanation
+
+### Role basics
+
+A role can be thought of as a group with any number of members belonging to the group.
+Roles have privileges on objects in the same way that users do. All members of the role
+inherit the privileges held by the role. This is done recursively.
+
+For example:
+* role `employees` has `ALL` privileges on table `mydb.employee_data`
+* user `marc` is a member of `employees` (also written: `marc ∈ employees`
+* without any extra privileges, `marc` can now perform any operations on table `mydb.employee_data`
+
+Roles and users follow a small set of rules:
+* roles and users use the same namespace (ie: if a user `marc` exists, we cannot create a role `marc`)
+* users and roles can be members of roles
+* membership loops are not allowed (direct: `A ∈ B ∈ A` or indirect: `A ∈ B ∈ C ... ∈ A`)
+* all privileges of a role are inherited by all its members
+* privileges on a table/database can be assigned to roles and users
+* only the admin role can create/drop roles
+* role admins (members with the `ADMIN OPTION`) can modify role memberships
+* a role cannot be dropped if it still has privileges
+* there is no limit (minimum or maximum) to the number of members of a role (except for the `admin` role)
+
+### Creating a role
+
+Cockroach administrators can create a new role called `myrole` using:
+```
+CREATE ROLE myrole
+```
+
+This fails if `myrole` exists either as a role or a user.
+
+### Granting and revoking privileges for a role
+
+Granting privileges to a role is the same a granting privileges to a user. This syntax has not changed.
+
+For example, granting `SELECT` privileges to the role `myrole` on a table consists of:
+```
+GRANT SELECT ON TABLE mydb.mytable TO myrole
+```
+
+Revoking privileges is also the same as before:
+```
+REVOKE SELECT ON TABLE mydb.mytable FROM myrole
+```
+
+### Manipulating role membership
+
+Modification of role membership requires one of the following:
+* cockroach administrator (`admin` role)
+* role admin (added to the role using `WITH ADMIN OPTION`)
+
+A user `marc` can be added to the role `myrole` without role admin permissions using:
+```
+GRANT myrole TO marc
+```
+
+`marc` is now a "regular" member of the role, inheriting all role privileges. However, `marc`
+is not allowed to modify role membership.
+
+To allow a user to modify role membership, we can use `WITH ADMIN OPTION`:
+```
+GRANT myrole TO marc WITH ADMIN OPTION
+```
+
+`marc` is now a member of the role **and** is allowed to modify role membership.
+
+### Dropping a role
+
+Cockroach administrators can delete a role called `myrole` using:
+```
+DROP ROLE myrole
+```
+
+This fails if `myrole` does not exist. We can instead use:
+```
+DROP ROLE IF EXISTS myrole
+```
+
+Dropping a role will fail if it still has privileges on database objects.
```

Comment:
Even indirect privileges via membership in another role?

Reply:
Nope, just privileges directly assigned. Memberships get automatically dropped but you need to clear out privileges first. Clarified the comment.